include/LightGBM/parser_base.h

```diff
@@ -0,0 +1,139 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#ifndef LIGHTGBM_PARSER_BASE_H_
+#define LIGHTGBM_PARSER_BASE_H_
+
+#include <functional>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace LightGBM {
+/*! \brief Interface for Parser */
+class Parser {
+ public:
+  typedef const char* (*AtofFunc)(const char* p, double* out);
+
+  /*! \brief Default constructor */
+  Parser() {}
+
+  /*!
+  * \brief Constructor for customized parser. The constructor accepts content not path because need to save/load the config along with model string
+  */
+  explicit Parser(std::string) {}
```

Comment:
to have a constructor of string looks like wield, not sure any purpose for this in base class?